core/system/src/ipc.c

```diff
@@ -313,8 +334,17 @@ void ipc_drain_queue(void)
          * of the descriptor, and clearing of the token. */
         uvisor_ipc_io_t * recv_io = &recv_array[recv_slot];
         if (!ipc_io_is_ok(recv_box_id, recv_io)) {
-            /* The IO is not entirely within the send box. Ignore it, and don't
-             * put it back. */
+            /* The receive IO is not entirely within the send box. Ignore it,
+             * and don't put it back. Put the send IO back because it is not
+             * malicious. */
+            put_it_back(send_queue, send_slot);
+            continue;
+        }
+
+        if (!ipc_io_token_is_ok(recv_ipc, recv_io)) {
+            /* The token of the receive IO is invalid. Ignore it, don't put it
+             * back. This shouldn't happen in a non-malicious box. Put the send
+             * IO back because it is not malicious. */
```

Comment:
I still do not understand something. Correct me if I am wrong:
The message is still not delivered, so the allocated_tokens on the receiving side is still not updated, which should cause ipc_io_token_is_ok() to fail here.
Is that correct?

Reply:
I don't understand what you mean. `allocated_tokens` will be updated by the user side.

Reply:
The aim of this fix is to prevent user assigning a token *maliciously* to the descriptor, so that the token system might be corrupted due to the malicious token.

Reply:
Example sequence of events:

Suppose there is a valid receive request 

Non-malicious:
- User: IPC send request
- uVisor API mode: allocate token, enqueue request, return control
- User: read token, possibly call wait functions on the tokens
- uVisor core mode: ipc_drain_queue: deliver message

Malicious:
- User: IPC send request
- uVisor API mode: allocate token, enqueue request, return control
- User: `desc->token = 0xffffffff`
- uVisor core mode: ipc_drain_queue: deliver message, `completed_tokens &= desc->token`
Now `completed_tokens` are corrupted

Reply:
Again, this will only shoot the malicious box in the foot, so this fix is optional.